qdax/core/emitters/mees_emitter.py

```diff
@@ -0,0 +1,613 @@
+            novelty_archive = jnp.zeros(
+                (self._total_generations, self._num_descriptors)
+            )
+        else:
+            novelty_archive = jnp.zeros(1)
+
+        # Create empty updated genotypes and fitness
+        last_updated_genotypes = jax.tree_map(
+            lambda x: jnp.zeros(shape=(5,) + x.shape[1:]),
+            init_genotypes,
+        )
+        last_updated_fitnesses = -jnp.inf * jnp.ones(shape=5)
+
+        return (
+            MEESEmitterState(
+                learning_rate=learning_rate,
+                initial_optimizer_state=initial_optimizer_state,
+                optimizer_state=initial_optimizer_state,
+                gradient_offspring=init_genotypes,
+                generation_count=0,
+                novelty_archive=novelty_archive,
+                last_updated_genotypes=last_updated_genotypes,
+                last_updated_fitnesses=last_updated_fitnesses,
+                last_updated_position=0,
+                random_key=random_key,
+            ),
+            random_key,
+        )
+
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
+    def emit(
+        self,
+        repertoire: MapElitesRepertoire,
+        emitter_state: MEESEmitterState,
+        random_key: RNGKey,
+    ) -> Tuple[Genotype, RNGKey]:
+        """Return the offspring generated through gradient update.
+
+        Params:
+            repertoire: the MAP-Elites repertoire to sample from
+            emitter_state
+            random_key: a jax PRNG random key
+
+        Returns:
+            a new gradient offspring
+            a new jax PRNG key
+        """
+
+        return emitter_state.gradient_offspring, random_key
+
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
+    def _novelty(
+        self,
+        generation_count: int,
+        descriptors: Descriptor,
+        novelty_archive: Descriptor,
+    ) -> jnp.ndarray:
+        """Compute the novelty of descriptors using a novelty archive.
+
+        Args:
+            generation_count: used to remove empty novelty archive slots
+            descriptors: array of descriptors of the individuals
+            novelty_archive: current novelty archive
+
+        Returns: the novelty score of all individuals
+        """
+
+        # Compute all distances with novelty_archive content
+        def distance(x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
+            return jnp.sqrt(jnp.sum(jnp.square(x - y)))
+
+        distances = jax.vmap(
+            jax.vmap(partial(distance), in_axes=(None, 0)), in_axes=(0, None)
+        )(descriptors, novelty_archive)
+
+        # Filter distance with empty slot of novelty archive
+        indices = jnp.arange(0, self._total_generations, step=1) < generation_count
+        distances = jax.vmap(lambda distance: jnp.where(indices, distance, jnp.inf))(
+            distances
+        )
+
+        # Find k neirest neighbours
```

Comment:
Have you experienced any time perf issue with the top_k? (btw, little typo in "nearest")

Reply:
I did not perform any tests. I was aware of the issue you raised on the speed of this function, but I did not see any other straightforward way to implement this here.